helix-rest/src/main/java/org/apache/helix/rest/server/service/InstanceServiceImpl.java

```diff
@@ -225,9 +248,11 @@ public StoppableCheck getInstanceStoppableCheck(String clusterId, String instanc
         StoppableCheck stoppableCheck = entry.getValue().get();
         if (!stoppableCheck.isStoppable()) {
           // put the check result of the failed-to-stop instances
-          finalStoppableCheckByInstance.put(instance, stoppableCheck);
-        } else {
-          // instance passed this around of check will be checked in the next round
+          addStoppableCheck(finalStoppableCheckByInstance, instance, stoppableCheck);
+        }
+        if (stoppableCheck.isStoppable() || _allChecks){
+          // instance passed this around of check or mandatory all checks
+          // will be checked in the next round
           instancesForNextCheck.add(instance);
         }
```

Comment:
They are correlated logic. Two "if" makes it independent checks.
So the logic should be: 

if (!stop && !_allChecks) {
   addStoppableCheck(finalStoppableCheckByInstance, instance, stoppableCheck);
} else {
   instancesForNextCheck.add(instance);
}

Reply:
@dasahcc I've thought about it and this logic won't work. Because when allChecks enabled, customers would like to have all the failed checks in the response. In the logic you presented, if allChecks == true, it won't add the failed check.

I do think the current 2 "if"s are clearer to understand.

Reply:
I see you concerns. Just last suggestion, we can do it in a separate RB. The logic for adding stoppable checks is not very good. We can keep a map of failed checks for instances as a global view instead of keeping creating StoppableCheck object. It was designed for old paths. Now it's time we can change it.

Reply:
Agreed. This was also something I would do to make it cleaner when working on the code. But it should be fine for now as one stoppable call does not have too many instances, so StoppableCheck objects won't be too many for the life cycle of a call. It'll need some refactoring work. An issue is created track it: #1692 .